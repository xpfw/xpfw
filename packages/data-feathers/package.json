{
  "name": "@xpfw/data-feathers",
  "version": "1.0.0",
  "main": "dist",
  "typings": "dist",
  "license": "MIT",
  "dependencies": {
    "@feathersjs/authentication-client": "^4.5.4",
    "@feathersjs/feathers": "^4.5.3",
    "@feathersjs/rest-client": "^4.5.4",
    "@feathersjs/socketio-client": "^4.5.4",
    "socket.io-client": "^2.3.0"
  },
  "peerDependencies": {
    "@xpfw/data": "1.x.x",
    "@xpfw/form": "1.x.x",
    "lodash": "4.x.x",
    "moment": "2.x.x"
  },
  "scripts": {
    "build": "./node_modules/.bin/rollup -c",
    "dev": "./node_modules/.bin/rollup -c -w",
    "test": "./node_modules/.bin/jest --ci",
    "update": "ncu -a",
    "prepare": "npm run build"
  },
  "devDependencies": {
    "@babel/runtime": "^7.9.6",
    "@types/feathersjs__authentication-client": "^1.0.5",
    "@types/feathersjs__feathers": "^3.1.5",
    "@types/feathersjs__rest-client": "^1.3.2",
    "@types/feathersjs__socketio-client": "^1.0.3",
    "@types/jest": "^25.2.1",
    "@types/lodash": "^4.14.150",
    "@types/lodash-es": "^4.17.3",
    "@types/mockdate": "^2.0.0",
    "@types/socket.io-client": "^1.4.32",
    "@xpfw/data": "^1.0.0",
    "@xpfw/data-tests": "1.0.0",
    "@xpfw/form": "^1.0.0",
    "@xpfw/test-util": "1.0.0",
    "ajv": "^6.12.2",
    "isomorphic-fetch": "^2.2.1",
    "jest": "^25.5.2",
    "jest-cli": "^25.5.2",
    "lodash": "^4.17.11",
    "lodash-es": "^4.17.15",
    "mobx": "^5.15.4",
    "mockdate": "^2.0.5",
    "moment": "^2.24.0",
    "react": "^16.13.1",
    "react-is": "^16.13.1",
    "rollup": "^2.7.5",
    "rollup-plugin-typescript2": "^0.27.0",
    "rollup-plugin-uglify": "^6.0.4",
    "ts-jest": "=25.4.0",
    "typescript": "^3.8.3",
    "uglify-es": "^3.3.10"
  },
  "jest": {
    "globals": {
      "ts-jest": {
        "diagnostics": false
      }
    },
    "transform": {
      ".(ts|tsx)": "ts-jest"
    },
    "testPathIgnorePatterns": [
      "\\.snap$",
      "<rootDir>/node_modules/",
      "<rootDir>/lib",
      "<rootDir>/dist"
    ],
    "moduleNameMapper": {
      "@xpfw/validate": "<rootDir>/node_modules/@xpfw/validate",
      "@xpfw/form": "<rootDir>/node_modules/@xpfw/form",
      "@xpfw/data": "<rootDir>/node_modules/@xpfw/data"
    },
    "moduleFileExtensions": [
      "ts",
      "tsx",
      "js"
    ],
    "testURL": "http://localhost/"
  }
}
