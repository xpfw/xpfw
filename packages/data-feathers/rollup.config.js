// rollup.config.js
import typescript from "rollup-plugin-typescript2"

export default {
  input: 'src/index.ts',
  output: {
    file: 'dist/index.js',
    format: 'cjs'
  },
  plugins: [
    typescript()
  ],
  external: [
    "isomorphic-fetch",
    "ajv",
    "lodash",
    "lodash-es",
    "mobx",
    "react",
    "react-dom",
    "preact-render-to-json",
    "moment",
    "mockdate",
    "@xpfw/form",
    "moment",
    "@feathersjs/feathers",
    "@feathersjs/authentication-client",
    "@feathersjs/socketio-client",
    "@feathersjs/rest-client",
    "resub",
    "resub-persist",
    "socket.io-client",
    "@xpfw/validate",
    "@xpfw/ui-tests",
    "@xpfw/ui-feathers",
    "@xpfw/data",
    "@xpfw/data-tests",
    "@xpfw/data-feathers",
    "@xpfw/test-util",
    "react",
    "react-dom",
    "isomorphic-fetch"
  ]
};