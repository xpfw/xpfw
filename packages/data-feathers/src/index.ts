import FeathersClient, { feathersClientOptions } from "./feathers"

export default FeathersClient
export {
  FeathersClient, feathersClientOptions
}
