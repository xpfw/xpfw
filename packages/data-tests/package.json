{
  "name": "@xpfw/data-tests",
  "version": "1.0.0-beta.16",
  "main": "dist",
  "typings": "dist",
  "license": "MIT",
  "peerDependencies": {
    "@xpfw/form": "1.x.x",
    "@xpfw/data": "1.x.x",
    "@xpfw/data-feathers": "1.x.x",
    "ajv": "^6.x.x",
    "lodash-es": "^4.x.x",
    "mobx": "^5.x.x",
    "react": "^16.x.x"
  },
  "scripts": {
    "build": "./node_modules/.bin/rollup -c",
    "dev": "./node_modules/.bin/rollup -c -w",
    "test": "./node_modules/.bin/jest --ci",
    "update": "ncu -a",
    "prepare": "npm run build"
  },
  "devDependencies": {
    "lodash": "^4.17.11",
    "lodash-es": "^4.17.11",
    "mobx": "^5.9.0",
    "ajv": "^6.10.0",
    "moment": "^2.24.0",
    "@types/jest": "^24.0.11",
    "@types/lodash": "^4.14.123",
    "@types/lodash-es": "^4.17.3",
    "@types/mockdate": "^2.0.0",
    "@types/react": "^16.8.8",
    "@types/socket.io-client": "^1.4.32",
    "@types/react-test-renderer": "^16.8.1",
    "@xpfw/test-util": "1.0.0-beta.16",
    "@xpfw/validate": "^0.1.2",
    "@xpfw/form": "^1.0.0-beta.16",
    "@xpfw/form-tests": "^1.0.0-beta.16",
    "@xpfw/form-shared": "^0.1.5",
    "@xpfw/data": "^1.0.0-beta.16",
    "@xpfw/data-feathers": "^1.0.0-beta.16",
    "jest": "^24.5.0",
    "jest-cli": "^24.5.0",
    "mockdate": "^2.0.2",
    "react": "^16.8.5",
    "react-test-renderer": "^16.8.5",
    "rollup": "^1.7.3",
    "rollup-plugin-typescript2": "^0.20.1",
    "rollup-plugin-uglify": "^6.0.2",
    "ts-jest": "=24.0.0",
    "tslib": "^1.9.3",
    "typescript": "^3.3.4000",
    "uglify-es": "^3.3.10"
  },
  "jest": {
    "globals": {
      "ts-jest": {
        "diagnostics": false
      }
    },
    "transform": {
      ".(ts|tsx)": "ts-jest"
    },
    "testPathIgnorePatterns": [
      "\\.snap$",
      "<rootDir>/node_modules/",
      "<rootDir>/lib",
      "<rootDir>/dist"
    ],
    "moduleNameMapper": {
    },
    "moduleFileExtensions": [
      "ts",
      "tsx",
      "js"
    ],
    "testURL": "http://localhost/"
  }
}
