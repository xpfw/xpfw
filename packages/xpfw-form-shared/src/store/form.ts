import {
  globals, IField, IFieldError, IForm, IFormError, iterateFields,
  validateField, validateForm, ValidationRegistry
} from "@xpfw/validate"
import { get, isFunction, isNil, isObject, set } from "lodash"
import { autoSubscribe, AutoSubscribeStore, key, StoreBase } from "resub"
import { IPersistableStore } from "resub-persist"
import FormErrorStore from "./error"

@AutoSubscribeStore
export class FormStore extends StoreBase implements IPersistableStore {
  public name = "formData"
  private formData: any = {}

  public getPropKeys() { return ["formData"] }

  public setValue(valuePath: string, value: any, fullTrigger: boolean = false) {
    set(this.formData, valuePath, value)
    this.trigger(fullTrigger ? undefined : valuePath)
  }

  @autoSubscribe
  public getValue(@key valuePath: string): any {
    return get(this.formData, valuePath)
  }

  public getFormData(form: IForm, prefix: string = "", method: string = "create") {
    const fullData: any = {}
    const formData = this.formData
    iterateFields(form, (field) => {
      let defaultValue = get(field, "validate.defaultValue")
      if (method === globals.Method.Find) {
        defaultValue = undefined
      }
      let value = get(formData,
        `${prefix && prefix.length > 0 ? prefix + "." : ""}${field.mapTo}`, defaultValue)
      let converter: any = get(field, "validate.convert")
      if (!isFunction(converter) && isObject(converter)) {
        converter = converter[method]
      }
      if (isFunction(converter)) {
        value = converter(value)
      }
      if (method !== globals.Method.Find || method === globals.Method.Find && !isNil(value))  {
        set(fullData, field.mapTo, value)
      }
    })
    return fullData
  }

  public copyDocToFormData(doc: any, form: IForm, prefix: string = "") {
    iterateFields(form, (field) => {
        this.setValue(`${prefix && prefix.length > 0 ? prefix + "." : ""}${field.mapTo}`, get(doc, field.mapTo))
    })
    this.trigger()
  }

  public async validateField(field: IField, form: IForm, method: string = "create",
                             prefix: string = "", currentUser?: string) {
    const valuePath = `${prefix && prefix.length > 0 ? prefix + "." : ""}${field.mapTo}`
    try {
      await validateField(get(this.formData, valuePath), field, {
        method,
        field,
        form,
        currentUser,
        registry: ValidationRegistry
      })
      FormErrorStore.setError(valuePath, {ok: true})
    } catch (e) {
      FormErrorStore.setError(valuePath, e)
    }
  }

  public async validateForm(form: IForm, method: string = "create", prefix: string = "", currentUser?: string) {
    const valuePath = `${prefix && prefix.length > 0 ? prefix + "." : ""}${form.model}`
    try {
      await validateForm(this.getFormData(form, prefix), form, {
        method,
        form,
        currentUser,
        registry: ValidationRegistry
      })
      FormErrorStore.setError(valuePath, {ok: true})
    } catch (e) {
      FormErrorStore.setError(valuePath, e)
    }
  }

  public resetForm(form: IForm, prefix: string = "")  {
    iterateFields(form, (field) => {
      this.setValue(`${prefix && prefix.length > 0 ? prefix + "." : ""}${field.mapTo}`, undefined)
    })
  }
}

export default new FormStore()
