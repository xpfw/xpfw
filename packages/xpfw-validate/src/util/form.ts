import { isObject } from "lodash"
import { IField, IForm } from "../typeDef"

const getFieldsFromForm: (form: IForm) => IField[] = (form: IForm) => {
  const fields: IField[] = []
  if (isObject(form)) {
    for (const section of form.sections) {
      for (const field of section.fields) {
        fields.push(field)
      }
    }
  }
  return fields
}

export { getFieldsFromForm }
