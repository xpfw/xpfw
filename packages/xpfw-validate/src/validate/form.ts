import { get, indexOf, intersection, isNil, isObject, isString, keys, set } from "lodash"
import { ErrorType, Method } from "../globals"
import getPermittedFields from "../permission/getPermittedFields"
import { IForm, IFormError, IParametersBase } from "../typeDef"
import validateField from "./any"
import validateQueryObject from "./queryObject"
import checkForFieldInSubKeys from "./util/checkForFieldInSubKeys"
import getRequireOwner from "./util/getRequireOwner"

const updateArrayOperators: string[] = [
  "$", "$inc", "addToSet", "$pop", "$pullAll", "$pull", "$pushAll", "$push",
  "$each", "$slice", "$sort", "$position", "$inc", "$mul", "$rename", "$setOnInsert",
  "$set", "$unset", "$min", "$max", "$currentDate", "$limit", "$skip", "$or", "$and"
]

const validateForm: (value: any, form: IForm, params: IParametersBase) => Promise<any> =
async (value, form, params) => {
  if (isObject(params)) {
    if (isNil(params.form)) {
      params.form = form
    }
    switch (params.method) {
      case Method.Find: {
        return validateQueryObject(value, form, params)
        break
      }
      case Method.Update: {
        try {
          const newRes = await checkForFieldInSubKeys(value, form, params, updateArrayOperators)
          return newRes
        } catch (e) {
          if (e !== -1) {
            throw e
          }
        }
        break
      }
    }
  }
  const validatedObject = {}
  const fields = await getPermittedFields(form, params, true)
  if (fields.length === 0) {
      return Promise.resolve(validatedObject)
  }
  const formError: IFormError = {
    model: form.model,
    errors: []
  }

  for (const field of fields) {
    try {
      const newVal = await validateField(get(value, field.mapTo), field, params)
      if (!isNil(newVal)) {
        set(validatedObject, field.mapTo, newVal)
      }
    } catch (e) {
      formError.errors.push(e)
    }
  }
  if (formError.errors.length > 0) {
    return Promise.reject(formError)
  }
  const method = isObject(params) && isString(params.method) ? params.method : ``
  if (method === Method.Update && keys(validatedObject).length === 0) {
    return Promise.reject({model: form.model, errors: [{
      model: `none`, type: ErrorType.emptyUpdate
    }]})
  }
  const requireOwner = getRequireOwner(form, method)
  if (isString(requireOwner)) {
    const idPath: any = params.idPath
    set(validatedObject, requireOwner, get(params.currentUser, idPath))
  }
  return Promise.resolve(validatedObject)
}

export default validateForm
