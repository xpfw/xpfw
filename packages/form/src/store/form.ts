import { get, set } from "lodash"
import { action, observable } from "mobx"
import { prependPrefix } from "../util/prefixMaker"

/**
 * Holds data relevant for Forms such as values and validation errors
 */
export class FormStoreClass {
  /**
   * MobX Observable containing all form data
   */
  @observable
  public formData: any = {}
  /**
   * MobX Observable containing all form errors
   */
  @observable
  public errors: any = {}
  /**
   * MobX Observable containing Loading states
   */
  @observable
  public loading: any = {}

  /**
   * Acquire a value from the form data
   * @param valuePath key where the value of the field will be saved. uses lodash.get/set string syntax style so subfields, array indices etc. can be used within the string
   * @param prefix prepended to mapTo to allow same mapTo keys to have different values
   * @param defaultValue value to return if it is not set in the form data
   */
  public getValue(valuePath?: string, prefix?: string, defaultValue?: any): any {
    return get(this.formData, prependPrefix(valuePath, prefix), defaultValue)
  }

  /**
   * Set a value in the form data
   * @param valuePath key where the value of the field will be saved. uses lodash.get/set string syntax style so subfields, array indices etc. can be used within the string
   * @param value the new value
   * @param prefix prepended to mapTo to allow same mapTo keys to have different values
   */
  @action
  public setValue(valuePath?: string, value?: any, prefix?: string) {
    set(this.formData, prependPrefix(valuePath, prefix), value)
  }

  /**
   * Acquire a value from the form errors
   * @param valuePath key where the value of the field will be saved. uses lodash.get/set string syntax style so subfields, array indices etc. can be used within the string
   * @param prefix prepended to mapTo to allow same mapTo keys to have different values
   */
  public getError(valuePath?: string, prefix?: string): any {
    return get(this.errors, prependPrefix(valuePath, prefix))
  }

  /**
   * Set a value in the form errors
   * @param valuePath key where the value of the field will be saved. uses lodash.get/set string syntax style so subfields, array indices etc. can be used within the string
   * @param value the new value
   * @param prefix prepended to mapTo to allow same mapTo keys to have different values
   */
  @action
  public setError(valuePath?: string, value?: any, prefix?: string) {
    set(this.errors, prependPrefix(valuePath, prefix), value)
  }

  /**
   * Acquire a value from the form errors
   * @param valuePath key where the value of the field will be saved. uses lodash.get/set string syntax style so subfields, array indices etc. can be used within the string
   * @param prefix prepended to mapTo to allow same mapTo keys to have different values
   */
  public getLoading(valuePath?: string, prefix?: string): any {
    return get(this.loading, prependPrefix(valuePath, prefix))
  }

  /**
   * Set a value in the form errors
   * @param valuePath key where the value of the field will be saved. uses lodash.get/set string syntax style so subfields, array indices etc. can be used within the string
   * @param value the new value
   * @param prefix prepended to mapTo to allow same mapTo keys to have different values
   */
  @action
  public setLoading(valuePath: string, value: boolean, prefix?: string) {
    set(this.loading, prependPrefix(valuePath, prefix), value)
  }
}

const FormStore = new FormStoreClass()
export default FormStore
