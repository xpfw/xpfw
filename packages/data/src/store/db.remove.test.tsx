import { FeathersClient } from "@xpfw/data-feathers"
import { toJS } from "@xpfw/data-tests"
import { FormStore } from "@xpfw/form"
import { makeSubFields, NameField, NumberAndRequiredTextSchema, NumberField } from "@xpfw/form-tests"
import { getRandomApp } from "@xpfw/test-util"
import "isomorphic-fetch"
import { get } from "lodash"
import BackendClient from "../client"
import DbStore from "./db"

BackendClient.client = FeathersClient

test("DbStore Remove Test", async () => {
  const s: any = NumberAndRequiredTextSchema.collection
  const appRef = await getRandomApp(s, false, BackendClient.client, true)
  expect(toJS(DbStore)).toMatchSnapshot("before anything")

  const fields = makeSubFields(NumberAndRequiredTextSchema)
  fields[NameField.title].setValue("myText")
  fields[NumberField.title].setValue(420)
  const createdObject = await DbStore.create(NumberAndRequiredTextSchema)
  expect(createdObject).toMatchSnapshot("createresult")
  const removeRes = await DbStore.remove(get(createdObject, "id"), NumberAndRequiredTextSchema.collection)
  expect(removeRes).toEqual(createdObject)
  expect(removeRes).toMatchSnapshot("removeresult")
  expect(toJS(DbStore)).toMatchSnapshot("After calling remove")
  const getResult = await DbStore.getFromServer(get(createdObject, "id"),
    NumberAndRequiredTextSchema.collection)
  expect(getResult).not.toBeNull()
  expect(getResult).not.toEqual(removeRes)
  expect(getResult).not.toEqual(createdObject)
  expect(removeRes).toEqual(DbStore.getRemoveState(get(createdObject, "id")))

  NumberAndRequiredTextSchema.collection = "broken"
  const res = await DbStore.remove(get(createdObject, "id"), NumberAndRequiredTextSchema.collection)
  expect(res).not.toBeNull()
  await appRef.cleanUp()
}, 10000)
