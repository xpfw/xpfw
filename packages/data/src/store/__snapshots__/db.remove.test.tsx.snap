// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DbStore Remove Test: After calling remove 1`] = `
Object {
  "createState": Object {
    "formModel": Object {
      "guestName": "myText",
      "id": 0,
      "myVal": 420,
    },
  },
  "currentlyEditing": "",
  "fetching": Object {},
  "getState": Object {
    "simpleTestCol": Object {},
  },
  "lastFetch": Object {},
  "removeState": Object {
    "0": Object {
      "guestName": "myText",
      "id": 0,
      "myVal": 420,
    },
  },
  "updateState": Object {},
}
`;

exports[`DbStore Remove Test: before anything 1`] = `
Object {
  "createState": Object {},
  "currentlyEditing": "",
  "fetching": Object {},
  "getState": Object {},
  "lastFetch": Object {},
  "removeState": Object {},
  "updateState": Object {},
}
`;

exports[`DbStore Remove Test: createresult 1`] = `
Object {
  "guestName": "myText",
  "id": 0,
  "myVal": 420,
}
`;

exports[`DbStore Remove Test: removeresult 1`] = `
Object {
  "guestName": "myText",
  "id": 0,
  "myVal": 420,
}
`;
